### PR TITLE
Ecmul Precompile

### PR DESCRIPTION
**Pull request by Recmo**

### CODE REVIEW

File: EIPS/eip-ecmul-precompile.md [hunk 1]
```diff
@@ -0,0 +1,151 @@
+---
+eip: To be assigned
+title: Precompile for Elliptic Curve Linear Combinations
+discussions-to: https://ethereum-magicians.org/t/ewasm-precompile-for-general-elliptic-curve-math/2581
+status: Draft
+type: Core
+replaces: EIP-196 (in some cases), EIP 665, EIP 1108 (partially), https://github.com/ethereum/EIPs/issues/603
```
Comment (axic) [inline EIPS/eip-ecmul-precompile.md]: Unfortunately this can only use an EIP number and that must be merged (cannot be an unmerged draft). For #603 I'd suggest you refer to it in a "References" or "Prior art" section.

File: EIPS/eip-ecmul-precompile.md [hunk 2]
```diff
@@ -0,0 +1,151 @@
+---
+eip: To be assigned
```
Comment (axic) [inline EIPS/eip-ecmul-precompile.md]: You can change this to 1829 and also rename the file.

File: EIPS/eip-ecmul-precompile.md [hunk 3]
```diff
@@ -0,0 +1,151 @@
...
+title: Precompile for Elliptic Curve Linear Combinations
+discussions-to: https://ethereum-magicians.org/t/ewasm-precompile-for-general-elliptic-curve-math/2581
+status: Draft
+type: Core
+replaces: EIP-196 (in some cases), EIP 665, EIP 1108 (partially), https://github.com/ethereum/EIPs/issues/603
+author: Remco Bloemen <Recmo@0x.org>
```
Comment (axic) [inline EIPS/eip-ecmul-precompile.md]: Could order these fields the same way the template has it?

File: EIPS/eip-ecmul-precompile.md [hunk 4]
```diff
@@ -0,0 +1,151 @@
...
+
+### Gas cost
+
+```
+BASE_GAS = ...
+ADD_GAS  = ...
+MUL_GAS  = ...
+```
+
+The total gas cost is `BASE_GAS` plus `ADD_GAS` for each $s_i$ that is $1$ and `MUL_GAS` for each $s_i > 1$ ($s_0 = 0$ is free).
```
Comment (axic) [inline EIPS/eip-ecmul-precompile.md]: Not sure markdown renders `$..$` but you could use backticks instead?

File: EIPS/eip-ecmul-precompile.md (hunk 1 above)
Comment (axic) [reply EIPS/eip-ecmul-precompile.md]: I actually don't think the `replaces` field is appropriate here. Perhaps it is as field not applicable to the EIP process we're following.

**Comment by Recmo** [comment]
@axic Thank you! I believe I have addressed all the comments.

**Comment by Recmo** [comment]
The Travis error does not seem related to the PR.

**Comment by vbuterin** [comment]
The thing that has always stopped me from proposing a similar thing is, what happens if someone submits a non-prime modulus? There are different ways to implement elliptic curve addition and multiplication, and in general I don't expect that the different libraries have been tested to ensure that they fail in the same cases when the modulus ends up being composite, and this seems like a large and deep potential source of consensus failures. Unfortunately primality tests that take into account weird edge cases like Carmichael numbers but are also clean and efficient don't really exist.

**Comment by Recmo** [comment]
@vbuterin So if the modulus is composite you end up with a Ring instead of a Field. The only difference being that inversion fails on more values than just zero.

Elliptic curve linear combinations, when implemented in the standard efficient way using Jacobi coordinates use only Ring operations. So the whole computation is safe. What will go wrong is converting back to Affine coordinates, which requires a single inversion. If this inversion fails, this would normally indicate the point at infinity. No new control flow. No edge cases being hit harder than otherwise, except for the last one.

Garbage in, garbage out still applies of course. Throw in a composite modulus and the thing you are computing is not guaranteed to be an elliptic curve.

Existing libraries do not seem particularly useful here. Most are designed for a particular curve. The remaining ones are mostly for research purposes and are not security hardened. (They would rely on probabilistic algorithms like Miller-Rabin or ).

The current implementation strategy is to develop a Rust library specific for this EIP which is generic and efficient, but only uses deterministic algorithms. This library will be available as Rust, C bindings and a eWASM module.

Currently the design allows you to use compressed coordinates as input. This gives developers some advantages, but requires us to compute modular square roots. Modular square root has similar issues as inversion. Simple deterministic procedures involving only Ring operations exist when `p = 3 mod 4` or `p = 5 mod 8`. Most field moduli used in practice satisfy this by design. My current thinking is to simply reject compressed coordinates when the field modulus does not satisfy and make compressed points optional.

Lastly, there's a feature request to do a form of hash-to-curve point. I think this can be incorporated elegantly and safely as well. I'll update the EIP soon to incorporate these ideas.
